### PR TITLE
Fix type inference for `on_retry`

### LINKED ISSUE #20: 0.4.0 regression: on_retry's type is now invalid under inference
**Describe the bug**

`on_retry` now requires an unsatisfiable lifetime.

**To Reproduce**

Try to compile the following:

```rust
use std::convert::Infallible;
use tryhard::retry_fn;

fn main() {
    let _ = async {
        let _ = retry_fn(|| async { Result::<_, Infallible>::Ok(()) })
            .retries(0)
            .on_retry(|_, _, _| async {})
            .await;
    };
}
```

This fails to compile with the error message:

```
    Updating crates.io index
    Checking tryhard-on-retry-regression v0.1.0 (/home/teo/junk/tryhard-on-retry-regression)
error: implementation of `std::ops::FnOnce` is not general enough
 --> src/main.rs:6:17
  |
6 |           let _ = retry_fn(|| async { Result::<_, Infallible>::Ok(()) })
  |  _________________^
7 | |             .retries(0)
8 | |             .on_retry(|_, _, _| async {})
9 | |             .await;
  | |__________________^ implementation of `std::ops::FnOnce` is not general enough
  |
  = note: closure with signature `fn(u32, std::option::Option<std::time::Duration>, &'2 std::convert::Infallible) -> impl std::future::Future` must implement `std::ops::FnOnce<(u32, std::option::Option<std::time::Duration>, &'1 std::convert::Infallible)>`, for any lifetime `'1`...
  = note: ...but it actually implements `std::ops::FnOnce<(u32, std::option::Option<std::time::Duration>, &'2 std::convert::Infallible)>`, for some specific lifetime `'2`

error: aborting due to previous error

error: could not compile `tryhard-on-retry-regression`
```

**Expected behavior**
The code should compile successfully, like it did on 0.3.0.

**Device:**
 - OS: Arch Linux
 - rustc 1.53.0 (53cb7b09b 2021-06-17) via Rustup

### PR DESCRIPTION
**Pull request by dvdplm**
Had a stab at #20. 

Still a bit hazy on why this change is needed but it fixes the problem afaict. Insipration came from reading through https://github.com/rust-lang/rust/issues/70263.

Q: what would a good way to test this be? Would a ui test be welcome, simply to prove that the code sample in #20 compiles?

Closes #20

### CODE REVIEW

**Review by davidpdrsn — APPROVED** [review]
Nice!

I guess adding

```rust
#[test]
fn this_compiles() {
    async {
        let _ = retry_fn(|| async { Result::<_, Infallible>::Ok(()) })
            .retries(0)
            .on_retry(|_, _, _| async {})
            .await;
    };
}
```

should do it.

File: src/lib.rs [hunk 1]
```diff
@@ -414,7 +414,10 @@ where
     /// # }
     /// ```
     #[inline]
-    pub fn on_retry<F>(self, f: F) -> RetryFuture<MakeFutureT, FutureT, BackoffT, F> {
+    pub fn on_retry<F, OnRetryFut>(self, f: F) -> RetryFuture<MakeFutureT, FutureT, BackoffT, F>
+    where
+        F: for<'a> Fn(u32, Option<Duration>, &E) -> OnRetryFut,
```
Comment (davidpdrsn) [inline src/lib.rs:419]: Do you actually need the `for<'a>`? `'a` doesn't seem to be referenced anywhere 🤔

Comment (dvdplm) [reply src/lib.rs]: You are absolutely right. Ty!

**Comment by clintfred** [comment]
I ran into this issue, but was able to work around it by giving a type annotation to the error.

@davidpdrsn Is there anything preventing this from being merged and a `0.4.1` from being released? Thanks!

**Comment by davidpdrsn** [comment]
@clintfred This needs a 0.5 since `on_retry` gained a new type parameter which is a breaking change.

Sorry for the huge delay on this. I'll publish this now :)

**Pull request merged**